File: python/2020/nov/numbers-at-most-n-given-digit-set.py
class Solution:
    def atMostNGivenDigitSet(self, digits, n):
        n = str(n)

        nLen = len(n)
        res = sum(len(digits) ** i for i in range(1, nLen))
        i = 0
        while i < len(n):
            res += sum(c < n[i] for c in digits) * (len(digits) ** (nLen - i - 1))
            if n[i] not in digits: break
            i += 1
        return res + (i == nLen)

File: archive/cpp/2021/dec/numbers-at-most-n-given-digit-set.cc
#include <string>
#include <vector>

using namespace std;

class Solution
{
public:
    int atMostNGivenDigitSet(vector<string> &digits, int n)
    {
        string ns = to_string(n);
        int digit = ns.size(), sz = digits.size(), res = 0;

        for (int i = 1; i < digit; ++i)
        {
            res += pow(sz, i);
        }
        for (int i = 0; i < digit; ++i)
        {
            bool has_same_num = false;
            for (string &d : digits)
            {
                if (d[0] < ns[i])
                {
                    res += pow(sz, digit - i - 1);
                }
                else if (d[0] == ns[i])
                {
                    has_same_num = true;
                }
            }
            if (!has_same_num)
            {
                return res;
            }
        }
        return res + 1;
    }
};
